src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -0,0 +1,338 @@
+            if (!string.IsNullOrEmpty(environmentPath))
+            {
+                _log.LogDebug($"Using ADB_EXE_PATH environment variable ({environmentPath}) for ADB path.");
+                adbExePath = environmentPath;
+            }
+            _absoluteAdbExePath = Path.GetFullPath(Environment.ExpandEnvironmentVariables(adbExePath));
+            if (!File.Exists(_absoluteAdbExePath))
+            {
+                _log.LogError($"Unable to find adb.exe");
+                throw new FileNotFoundException($"Could not find adb.exe. Either set it in the environment via ADB_EXE_PATH or call with valid path (provided:  '{adbExePath}')");
+            }
+            if (!_absoluteAdbExePath.Equals(adbExePath))
+            {
+                _log.LogDebug($"ADBRunner using ADB.exe supplied from {adbExePath}");
+                _log.LogDebug($"Full resolved path:'{_absoluteAdbExePath}'");
+            }
+        }
+
+        #endregion
+
+        #region Functions
+
+        public string GetAdbVersion()
+        {
+            return RunAdbCommand("version").StandardOutput;
+        }
+
+        public string GetAdbState()
+        {
+            return RunAdbCommand("get-state").StandardOutput;
+        }
+
+        public void ClearAdbLog()
+        {
+            RunAdbCommand("logcat -c");
+        }
+
+        public void DumpAdbLog(string outputFilePath, string filterSpec = "")
+        {
+            // This is somehow reaching full timeout, have to figure it out, allow 20s for demo?
+            var result = RunAdbCommand($"logcat -d {filterSpec}");
+            if (result.ExitCode != 0)
+            {
+                // Could throw here, but it would tear down a possibly otherwise acceptable execution.
+                _log.LogError($"Error getting ADB log:{Environment.NewLine}{FormatProcessOutputs(result)}");
+            }
+            else
+            {
+                Directory.CreateDirectory(Path.GetDirectoryName(outputFilePath));
+                File.WriteAllText(outputFilePath, result.StandardOutput);
+                _log.LogInformation($"Wrote current ADB log to {outputFilePath}");
+            }
+        }
+
+        public void StartAdbServer()
+        {
+            var result = RunAdbCommand("start-server");
+            _log.LogDebug($"{result.StandardOutput}");
+            if (result.ExitCode != 0)
+            {
+                throw new Exception($"Error starting ADB Server.  Std out:{result.StandardOutput} Std. Err: {result.StandardError}");
+            }
+        }
+
+        public void KillAdbServer()
+        {
+            var result = RunAdbCommand("kill-server");
+            if (result.ExitCode != 0)
+            {
+                throw new Exception($"Error killing ADB Server.  Std out:{result.StandardOutput} Std. Err: {result.StandardError}");
+            }
+        }
+
+        public int InstallApk(string apkPath)
+        {

```

Comment:
Can appPath be NullOrEmty? Also, could the path be missing? We might have an installation failure because the path is not present, and error stating that it could not find the app would give more info than an install error.